Lib/test/test_asyncio/test_timeouts.py

```diff
@@ -0,0 +1,221 @@
+        self.assertTrue(cm.expired())
+
+    async def test_timeout_at_basic(self):
+        loop = asyncio.get_running_loop()
+
+        with self.assertRaises(TimeoutError):
+            deadline = loop.time() + 0.01
+            async with asyncio.timeout_at(deadline) as cm:
+                await asyncio.sleep(10)
+        self.assertTrue(cm.expired())
+        self.assertEqual(deadline, cm.when())
+
+    async def test_nested_timeouts(self):
+        cancel = False
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.01) as cm1:
+                # Only the topmost context manager should raise TimeoutError
+                with self.assertRaises(asyncio.CancelledError):
+                    async with asyncio.timeout(0.01) as cm2:
+                        await asyncio.sleep(10)
+        self.assertTrue(cm1.expired())
+        self.assertTrue(cm2.expired())
+
+    async def test_waiter_cancelled(self):
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.01):
+                with self.assertRaises(asyncio.CancelledError):
+                    await asyncio.sleep(10)
+
+    async def test_timeout_not_called(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout(10) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+        self.assertGreater(cm.when(), t1)
+
+    async def test_timeout_disabled(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout(None) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        self.assertIsNone(cm.when())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+
+    async def test_timeout_at_disabled(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout_at(None) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        self.assertIsNone(cm.when())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+
+    async def test_timeout_zero(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0) as cm:
+                await asyncio.sleep(10)
+        t1 = loop.time()
+        self.assertTrue(cm.expired())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+        self.assertTrue(t0 <= cm.when() <= t1)
+
+    async def test_foreign_exception_passed(self):
+        with self.assertRaises(KeyError):
+            async with asyncio.timeout(0.01) as cm:
+                raise KeyError
+        self.assertFalse(cm.expired())
+
```

Comment:
Here's a test for the foreign exception upon timeout.
```suggestion

        async def test_foreign_exception_on_timeout(self):
        async def crash():
            try:
                await asyncio.sleep(1)
            finally:
                1/0
        with self.assertRaises(ZeroDivisionError):
            async with asyncio.timeout(0.01):
                await crash()
```